Provide CLI for all Nomadize commands

moonglum opened the pull request:
An alternative way to call the Nomadize commands for people that want to call it directly or for example via make.

Not everyone uses Rake as their task runner, some people use Make instead. This executable makes it possible to use Nomadize via Make. It also allows it to just run it via the commandline of course :smile:

moonglum committed:
Provide CLI for all Nomadize commands

An alternative way to call the Nomadize commands for people that want to
call it directly or for example via make.

piisalie commented:
This is an excellent addition. Thank you @moonglum!

One question: Do we want to include a `-h` option? `$command -h` is the first thing I try when I encounter an error or unexpected behavior using a command line tool.

The downside I see is having to maintain a list of commands and explanations in two places ( in the README and in the command line tool ). Thoughts?

moonglum commented:
Hey @piisalie – glad you like it :smile: I thought about that, too. Currently the information is in two places (the README and the `desc`s in the Rakefile). That would add a third place with that knowledge... It would be possible to keep it at two places by keeping the descriptions in a central place (something like a hash) and reading that information from both `tasks.rb` as well as the CLI. But I'm not sure if that's a good idea :wink:

piisalie commented:
You're completely right, it would be three places; I didn't think about the Rake task descriptions. :smile: 

I am also unsure about using a big hash. In theory this library shouldn't be changing that much so maybe having the descriptions in two places in the code and one place in the docs is acceptable duplication.

Also note, I am a single person mentioning the `-h` flag. It could totally be that this is not something anyone else cares about and we could move on without it for now. ;-)

moonglum commented:
Something like this could work:

``` ruby
module Nomadize
  @help = {
    create: 'Create database using name in {appdir}/config/database.yml'
  }
  class << self; attr_reader :help; end
end
```

This file is then required by tasks.rb:

``` ruby
desc Nomadize.help.fetch(:create)
task :create do
  Nomadize.create_database
end
```

and by the CLI:

``` ruby
puts "Usage:"
Nomadize.help.each_pair do |command, help|
  puts "  nomadize #{command.to_s.ljust(10)} #{help}"
end
```

moonglum commented:
I think `-h` is pretty common and should be supported :smile: I just don't know which variant is best...

piisalie commented:
I like your hash solution; it is simple and works in both cases.

moonglum commented:
Okidoke, I'll add it to this PR as soon as I have the time :smile:

piisalie commented:
Thanks so much! I look forward to utilizing the shiny green merge button. ;-)

moonglum committed:
Extract help information from Rake Tasks

This is in preparation to use it for the CLI as well

moonglum committed:
Add -h and --help option to the CLI

moonglum commented:
Done! Rebased it onto the new master, also updated the `new_migration` help accordingly :wink:

btw.: Thanks for the great project :smile: I'm currently using it for a toy project where I try out a few libraries (mainly [tynn](http://tynn.xyz), sql_capsule and nomadize) and really like it so far :smile:

piisalie commented:
Woot! :+1: Thanks for your pull request and your kind words. :smile: 

I hadn't seen tynn, though I was just recently playing with Syro. I'll have to give it a try.

moonglum commented:
You're more than welcome :smile: 

Check it out :smile: If you like Syro, it will probably be interesting for you as it is Syro + some things you need most of the time (template rendering, secure headers etc.) :smile:

piisalie commented:
I've only tinkered with Syro thus far, most of my toy projects have used Cuba. Tynn does look like it would be super helpful. I quite like the idea of building small abstractions / libraries on top of other small libraries / abstractions.

Also, version 0.3.0 should be released to RubyGems; new and improved with command line support!!! :smile:

moonglum commented:
Yep, I like Cuba :smile: And Syro is a very good new iteration on Cuba :smile:

Thank you so much :heart: Now I can switch my tiny project over to Make :dancers:
